### PR TITLE
vegeta plot command

### LINKED ISSUE #137: Plot reporter time series compression
For very large result sets, the plot reporter generates a gigantic HTML+JS file which is very very slow to render in a browser. We can be smarter here and apply some compression to the result set without loosing most information needed.

https://www.cs.cmu.edu/~eugene/research/full/compress-series.pdf
http://skemman.is/stream/get/1946/15343/37285/3/SS_MSthesis.pdf

### PR DESCRIPTION
**Pull request by tsenart**
This PR introduces a new `vegeta plot` command that is an evolution of the previous `vegeta report -reporter=plot`. Additionally, all plot related code has moved to its own package.

### Features

- Support for plotting very large result sets with bounded memory usage.
- Downsampling via LTTB algorithm, used via a flag in the command.
- Plotting multiple attacks overlaid in the same line chart.

Fixes #137 

### TODO
- [x] Golden test for plot output
- [x] Update Dygraphs library to latest version
- [x] Fork https://github.com/dgryski/go-tsz to use uint64 timestamps.
- [x] Research and implement improvements to Gorilla as per https://github.com/burmanm/gorilla-tsc.

### Future work

-  Real-time plotting with a web sockets / SSE / HTTP2 Push + JS integration.

### CODE REVIEW

File: lib/reporters.go [hunk 1]
```diff
@@ -109,51 +111,70 @@ func NewJSONReporter(m *Metrics) Reporter {
...
-		for _, r := range *rs {
-			attacks[r.Attack] = append(attacks[r.Attack], r)
+		samples := make(map[string][][]lttb.Point, len(series))
+		for attack, results := range series {
+			for i := range results {
+				sort.Sort(results[i])
+				points := make([]lttb.Point, 0, len(results[i]))
+				for _, r := range results[i] {
+					points = append(points, lttb.Point{
+						X: float64(r.Timestamp.Sub(results[i][0].Timestamp).Seconds()),
```
Comment (golangci) [inline lib/reporters.go]: unnecessary conversion

File: lib/reporters.go [hunk 2]
```diff
@@ -109,51 +111,70 @@ func NewJSONReporter(m *Metrics) Reporter {
...
+						Y: float64(r.Latency.Seconds() * 1000),
```
Comment (golangci) [inline lib/reporters.go]: unnecessary conversion

**Comment by tsenart** [comment]
Here's a visual representation of the effect of this change set. As for the performance effects, you can test by downloading the attached HTML files and opening each of them.

[Plot Reports](https://github.com/tsenart/vegeta/files/2193568/plot-reports.zip)

### Before
vegeta-plot-raw

### After
vegeta-plot-downsampled

File: lib/plot_test.go [hunk 3]
```diff
@@ -20,11 +20,10 @@ func BenchmarkPlotReporter(b *testing.B) {
 			rs[i].Error = "Error"
 		}
 	}
-	rep := NewPlotReporter("Vegeta Plot", &rs)
 	// Start benchmark
 	b.ReportAllocs()
 	b.StartTimer()
 	for i := 0; i < b.N; i++ {
-		rep.Report(ioutil.Discard)
+		NewHTMLPlot("Vegeta Plot", 5000, rs).WriteTo(ioutil.Discard)
```
Comment (golangci) [inline lib/plot_test.go]: Error return value of `(github.com/tsenart/vegeta/lib.HTMLPlot).WriteTo` is not checked

File: plot.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	fs := flag.NewFlagSet("vegeta plot", flag.ExitOnError)
+	title := fs.String("title", "Vegeta Plot", "Title and header of the resulting HTML page")
+	threshold := fs.Int("threshold", 4000, "Threshold of data points above which series are downsampled.")
+	output := fs.String("output", "stdout", "Output file")
+
+	fs.Usage = func() {
+		fmt.Fprintln(os.Stderr, plotUsage)
+	}
+
+	return command{fs, func(args []string) error {
```
Comment (golangci) [inline plot.go]: github.com/tsenart/vegeta.command composite literal uses unkeyed fields

File: plot.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+		fs.Parse(args)
```
Comment (golangci) [inline plot.go]: Error return value of `fs.Parse` is not checked

File: lib/plot_test.go [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+		}
+		if i%5 == 0 {
+			rs[i].Error = "Error"
+		}
+	}
+	// Start benchmark
+	b.ReportAllocs()
+	b.StartTimer()
+	for i := 0; i < b.N; i++ {
+		NewHTMLPlot("Vegeta Plot", 5000, rs).WriteTo(ioutil.Discard)
```
Comment (golangci) [inline lib/plot_test.go]: Error return value of `(github.com/tsenart/vegeta/lib.HTMLPlot).WriteTo` is not checked

File: lib/plot.go [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+	optsJSON, err := json.MarshalIndent(&opts, "    ", " ")
+	if err != nil {
+		return 0, err
+	}
+
+	cw := countingWriter{w: w}
+	err = plotTemplate.Execute(&cw, &plotData{
+		Title:         p.title,
+		HTML2CanvasJS: template.JS(asset(html2canvas)),
```
Comment (golangci) [inline lib/plot.go]: G203: this method will not auto-escape HTML. Verify data is well formed.

File: lib/plot.go [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+		DygraphsJS:    template.JS(asset(dygraphs)),
```
Comment (golangci) [inline lib/plot.go]: G203: this method will not auto-escape HTML. Verify data is well formed.

File: lib/plot.go [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+		Data:          template.JS(data),
```
Comment (golangci) [inline lib/plot.go]: G203: this method will not auto-escape HTML. Verify data is well formed.

File: lib/lttb_test.go [hunk 10]
```diff
@@ -0,0 +1,14102 @@
...
+		ps := data[:count]
+		data = data[count:]
+		return ps, nil
+	}
+}
+
+// From https://raw.githubusercontent.com/sveinn-steinarsson/flot-downsample/master/demo_data.js
+var points = [][]point{
+	{
+		{0, 29.357995947822218},
```
Comment (golangci) [inline lib/lttb_test.go]: github.com/tsenart/vegeta/lib.point composite literal uses unkeyed fields

File: lib/lttb_test.go [hunk 11]
```diff
@@ -0,0 +1,14102 @@
...
+		{1, 29.40932479606209},
```
Comment (golangci) [inline lib/lttb_test.go]: github.com/tsenart/vegeta/lib.point composite literal uses unkeyed fields

File: lib/lttb_test.go [hunk 12]
```diff
@@ -0,0 +1,14102 @@
...
+		{2, 29.28168582006162},
```
Comment (golangci) [inline lib/lttb_test.go]: github.com/tsenart/vegeta/lib.point composite literal uses unkeyed fields

File: lib/lttb_test.go [hunk 13]
```diff
@@ -0,0 +1,14102 @@
...
+	data := *(*[]golttb.Point)(unsafe.Pointer(&points[0]))
+	b.Run("dgryski", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			golttb.LTTB(data, 1000)
+		}
+	})
+
+	b.Run("tsenart", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			lttb(len(data), 1000, newIterator(points[0]))
```
Comment (golangci) [inline lib/lttb_test.go]: Error return value of `lttb` is not checked

File: lib/lttb/lttb_test.go [hunk 14]
```diff
@@ -0,0 +1,14100 @@
...
+	data := *(*[]golttb.Point)(unsafe.Pointer(&points[0]))
+	b.Run("dgryski", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			golttb.LTTB(data, 1000)
+		}
+	})
+
+	b.Run("tsenart", func(b *testing.B) {
+		for i := 0; i < b.N; i++ {
+			Downsample(len(data), 1000, newIterator(points[0]))
```
Comment (golangci) [inline lib/lttb/lttb_test.go]: Error return value of `Downsample` is not checked

**Pull request merged**